src/lib/syncPagePath.ts

```diff
@@ -0,0 +1,413 @@
+      if (!regex.test(payload.parent.uri)) {
+        let uri = payload.parent.uri;
+        this.debug(`normalizing parent uri ${uri} for ${filePath}`);
+        // remove leading and trailing slashes
+        uri = uri.replace(/^\/|\/$/g, '');
+        payload.parent.uri = `/versions/${version}/${this.route}/${uri}`;
+      }
+    }
+
+    const createPage = (): Promise<PushResult> | PushResult => {
+      if (dryRun) {
+        return { filePath, result: 'created', response: null, slug };
+      }
+
+      return this.readmeAPIFetch(
+        route,
+        { method: 'POST', headers, body: JSON.stringify(payload) },
+        { filePath, fileType: 'path' },
+      )
+        .then(res => this.handleAPIRes(res))
+        .then(res => {
+          return { filePath, result: 'created', response: res, slug };
+        });
+    };
+
+    const updatePage = (): Promise<PushResult> | PushResult => {
+      if (dryRun) {
+        return { filePath, result: 'updated', response: null, slug };
+      }
+
+      // omits the slug from the PATCH payload since this would lead to unexpected behavior
+      delete payload.slug;
+
+      return this.readmeAPIFetch(
+        `${route}/${slug}`,
+        {
+          method: 'PATCH',
+          headers,
+          body: JSON.stringify(payload),
+        },
+        { filePath, fileType: 'path' },
+      )
+        .then(res => this.handleAPIRes(res))
+        .then(res => {
+          return { filePath, result: 'updated', response: res, slug };
+        });
+    };
+
+    return this.readmeAPIFetch(`${route}/${slug}`, {
+      method: 'HEAD',
+      headers,
+    }).then(async res => {
+      await this.handleAPIRes(res, true);
+      if (!res.ok) {
+        if (res.status !== 404) throw new APIv2Error(res);
+        this.debug(`error retrieving data for ${slug}, creating page`);
+        return createPage();
+      }
+      this.debug(`data received for ${slug}, updating page`);
+      return updatePage();
+    });
+  } catch (e) {
+    return { error: e, filePath, result: 'failed', slug };
+  }
+}
+
+const byParentPage = (left: PageMetadata<PageRepresentation>, right: PageMetadata<PageRepresentation>) => {
+  return (right.data.parent?.uri ? 1 : 0) - (left.data.parent?.uri ? 1 : 0);
+};
+
+function sortFiles(files: PageMetadata<PageRepresentation>[]): PageMetadata<PageRepresentation>[] {
+  const filesBySlug = files.reduce<Record<string, PageMetadata<PageRepresentation>>>((bySlug, obj) => {
+    // eslint-disable-next-line no-param-reassign
+    bySlug[obj.slug] = obj;
+    return bySlug;
+  }, {});
+  const dependencies = Object.values(filesBySlug).reduce<
+    [PageMetadata<PageRepresentation>, PageMetadata<PageRepresentation>][]
+  >((edges, obj) => {
+    if (obj.data.parent?.uri && filesBySlug[obj.data.parent.uri]) {
+      edges.push([filesBySlug[obj.data.parent.uri], filesBySlug[obj.slug]]);
+    }
+
+    return edges;
+  }, []);
+
+  return toposort.array(files, dependencies);
```

Comment:
i follow what `toposort` is doing from their docs, but how come it's being used here? to ensure that we create parents before children?

mind adding some comments?

Reply:
can do! for the record this is carried over from @kellyjosephprice's work over in https://github.com/readmeio/rdme/pull/973 and i've blindly trusted it ever since lol

Reply:
https://github.com/readmeio/rdme/pull/1159/commits/8153b90bf512eb65b4a8338fc0445a4e28796f72